lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,53 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||=  read_through { |obj_type| @schema.union_memberships(obj_type) }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
+        return false unless visible?(type_defn)
+
+        if type_defn.kind.object? && has_abstract_type?(type_defn)
+          visible_implementation?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn)
+        end
+      end
+
+      def visible_implementation?(type_defn)
+        if orphan?(type_defn)
```

Comment:
It's _possible_ that a type would be: 

- Specified as an orphan _AND_ 
- Referenced somewhere in the schema 

I don't think graphql-ruby disallows that. For example, it might happen if a type _used to_ be an orphan but was eventually used somewhere else in the schema. 

In that case, we should still check for references to the type, right? If so, then we should always use the `else` branch for this method, so the whole method body would be: 

```ruby
referenced_by_visible_members?(type_defn) || member_or_implements?(type_defn)
```